internal/users/manager.go

```diff
@@ -197,20 +215,31 @@ func (m *Manager) UpdateUser(u types.UserInfo) (err error) {
 		}
 
 		if g.GID == nil {
-			// The group does not exist in the database, so we generate a unique GID for it. Similar to the RegisterUser
-			// call above, this also registers a temporary group in our NSS handler. We remove that temporary group
-			// before returning from this function, at which point the group is added to the database (so we don't need
-			// the temporary group anymore to keep the GID unique).
-			gid, cleanup, err := m.temporaryRecords.RegisterGroup(g.Name)
+			// The group does not exist in the database, so we generate a unique GID for it.
+			newGroups = append(newGroups, g)
+			continue
+		}
+
+		groupRows = append(groupRows, db.NewGroupRow(g.Name, *g.GID, g.UGID))
+	}
+
+	if len(newGroups) > 0 {
+		localEntries, unlockEntries, err := localentries.NewWithLock()
+		if err != nil {
+			return err
+		}
+		defer func() { err = errors.Join(err, unlockEntries()) }()
+		records := m.temporaryRecords.LockForChanges(localEntries)
+
+		for _, g := range newGroups {
+			gid, cleanup, err := records.RegisterGroupForUser(uid, g.Name)
```

Comment:
Same here, we don't need to register the group if we just hold a lock instead.

Reply:
Partially, given that for each group we need also to ensure that it is not clashing with other groups and with the current user UID